File: src/stringparser/StringParser.java
/*
 * Copyright (c) 2015 Julien Guerinet. All rights reserved.
 */

package stringparser;

import org.supercsv.cellprocessor.ift.CellProcessor;
import org.supercsv.io.CsvListReader;
import org.supercsv.prefs.CsvPreference;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

/**
 * Main class, executes the main code for parsing the Google Docs file
 * @author Julien Guerinet
 * @version 1.0
 * @since 1.0
 */
public class StringParser{
    /* PLATFORMS */
    /**
     * The platform chosen
     */
    private static int platform = -1;
    /**
     * Android platform
     */
    private static final int ANDROID = 0;
    /**
     * iOS platform
     */
    private static final int IOS = 1;

    /* FILE STRINGS */
    /**
     * The URL in the file
     */
    private static final String URL = "URL:";
    /**
     * The platform in the file
     */
    private static final String PLATFORM = "Platform:";
    /**
     * Languages in the file
     */
    private static final String LANGUAGE= "Language:";

    /* ANDROID STRINGS */
    /**
     * Android XML Opener
     */
    private static final String XML_OPENER = "<?xml version=\"1.0\" encoding=\"utf-8\"?>";
    /**
     * Android Resources Opener
     */
    private static final String RESOURCES_OPENER = "<resources>";
    /**
     * Android Resources Closer
     */
    private static final String RESOURCES_CLOSER = "</resources>";

    /* OTHER */
    /**
     * The English Id, used to get the English Strings for the header
     */
    private static final String ENGLISH_ID = "en";
    /**
     * The key used for the header in the Strings document
     */
    private static final String HEADER_KEY = "header";

    public static void main(String[] args) throws IOException {
        //Keep a list of all of the languages the Strings are in
        List<Language> languages = new ArrayList<Language>();
        //The list of language Strings
        List<LanguageString> strings = new ArrayList<LanguageString>();
        //Url
        String url = null;

        //Read from the config file
        BufferedReader configReader = null;
        try{
            configReader = new BufferedReader(new FileReader("../config.txt"));
        }
        catch(FileNotFoundException e){
            try{
                configReader = new BufferedReader(new FileReader("config.txt"));
            }
            catch(FileNotFoundException ex){
                System.out.println("Error: Config file not found");
                System.exit(-1);
            }
        }

        String line;
        while ((line = configReader.readLine()) != null) {
            //Get the URL
            if(line.startsWith(URL)){
                url = line.replace(URL, "").trim();
            }
            //Get the platform
            else if(line.startsWith(PLATFORM)){
                //Remove the header
                String platformString = line.replace(PLATFORM, "").trim();
                //Android
                if(platformString.equalsIgnoreCase("android")){
                    platform = ANDROID;
                }
                //iOS
                else if(platformString.equalsIgnoreCase("ios")){
                    platform = IOS;
                }
                //Not recognized
                else{
                    System.out.println("Error: Platform must be either Android, iOS, or Windows.");
                }
            }
            //Get the languages
            else if(line.startsWith(LANGUAGE)){
                //Remove the header and separate the language Id from the path
                String[] languageInfo = line.replace(LANGUAGE, "").trim().split(", ");

                //Save it as a new language in the list of languages
                languages.add(new Language(languageInfo[0], languageInfo[1]));
            }
        }
        configReader.close();

        //Make sure nothing is null
        if(url == null){
            System.out.println("Error: URL Cannot be null");
            System.exit(-1);
        }
        else if(platform == -1){
            System.out.println("Error: You need to input a platform");
            System.exit(-1);
        }
        else if(languages.isEmpty()){
            System.out.println("Error: You need to add at least one language");
            System.exit(-1);
        }

        //Connect to the URL
        URL link = new URL(url);
        HttpURLConnection connection = (HttpURLConnection) link.openConnection();
        connection.setRequestMethod("GET");
        connection.connect();

        //Set up the CSV reader
        CsvListReader reader = new CsvListReader(new InputStreamReader(connection.getInputStream()),
                CsvPreference.EXCEL_PREFERENCE);

        //Get the header
        final String[] header = reader.getHeader(true);

        //First column will be key, so ignore it
        for(int i = 1; i < header.length; i++){
            String string = header[i];

            //Check if the string matches any of the languages parsed
            for(Language language : languages){
                if(string.equals(language.getId())){
                    //If we find a match, set the column index for this language
                    language.setColumnIndex(i);
                    break;
                }
            }
        }

        //Make sure that all languages have an index
        for(Language language : languages){
            if(language.getColumnIndex() == -1){
                System.out.println("Error: " + language.getId() +
                        " does not have any translations.");
                System.exit(-1);
            }
        }

        //Make a CellProcessor with the right length
        final CellProcessor[] processors = new CellProcessor[header.length];

        //Go through each line of the CSV document into a list of objects.
        List<Object> currentLine;
        while((currentLine = reader.read(processors)) != null){
            //Add a new language String
            LanguageString languageString = new LanguageString((String)currentLine.get(0));

            //Go through the languages, add each translation
            boolean allNull = true;
            for(Language language : languages){
                languageString.addTranslation(language.getId(),
                        (String)currentLine.get(language.getColumnIndex()));

                //If at least one language is not null, then they are not all null
                if(languageString.getString(language.getId()) != null){
                    allNull = false;
                }
            }

            //Check if all of the values are null
            if(allNull){
                //Show a warning message
                System.out.println("Warning: Line " + (strings.size() + 2) + " has no " +
                    "translations so it will not be parsed.");
            }
            else{
                strings.add(languageString);
            }
        }

        //Close the CSV reader
        reader.close();

        //Check if there are any errors with the keys
        for (int i = 0; i < strings.size(); i++){
            LanguageString string1 = strings.get(i);

            //Check if there are any spaces in the keys
            if(string1.getKey().trim().contains(" ")){
                System.out.println("Error: Line " + getLineNumber(string1, strings) + " contains " +
                                "a space in its key.");
                System.exit(-1);
            }

            //Check if there are any duplicated
            for(int j = i + 1; j < strings.size(); j++){
                LanguageString string2 = strings.get(j);

                //If the keys are the same and it's not a header, show an error and stop
                if(!string1.getKey().equalsIgnoreCase(HEADER_KEY) &&
                        string1.getKey().equals(string2.getKey())){
                    System.out.println("Error: Lines " + getLineNumber(string1, strings) + " and " +
                            getLineNumber(string2, strings) + " have the same key.");
                    System.exit(-1);
                }
            }
        }

        //Go through each language, and write the file
        for(Language language : languages){
            if(platform == ANDROID){
                processAndroidStrings(language, strings);
            }
            else{
                processIOSStrings(language, strings);
            }
        }

        //Exit message
        System.out.println("Done parsing Strings");
    }

    /* HELPERS */

    /**
     * Get the line number of a given String for any warnings or errors shown to the user
     *
     * @param string The String
     * @return The line number of the String
     */
    private static int getLineNumber(LanguageString string, List<LanguageString> strings){
        //+2 to account for the header and the fact that Google Drive starts numbering at 1
        return strings.indexOf(string) + 2;
    }

    /**
     * Processes a given String with the common changes to make between the platforms
     *
     * @param string The String to process
     */
    private static String processString(String string){
        //Unescaped quotations
        string = string.replace("\"", "\\" + "\"");

        //Copyright
        string = string.replace("(c)", "\u00A9");

        return string;
    }

    /**
     * Add a language String
     *
     * @param platform The platform this is for
     * @param string   The LanguageString object
     * @param language The language to parse the String for
     * @return The formatted String for the given language and platform
     */
    private static String getLanguageString(int platform, LanguageString string, Language language){
        String key = string.getKey();
        String value;
        //Check if we are parsing a header, use the English translation for the value
        if(string.getKey().equalsIgnoreCase(HEADER_KEY)){
            value = string.getString(ENGLISH_ID);
        }
        else{
            value = string.getString(language.getId());
        }

        //Check if value is or null empty: if it is, return null
        if(value == null || value.isEmpty()){
            return null;
        }

        //Process the value with the general methods first
        value = processString(value);

        //Use the right platform method
        if(platform == ANDROID){
            return getAndroidString(key, value);
        }
        return getIOSString(key, value);
    }

    /* ANDROID STRING PARSING */

    /**
     * Get the formatted String for the Android Strings document
     *
     * @param key    The String key
     * @param string The String
     * @return The formatted String
     */
    private static String getAndroidString(String key, String string){
        //Add initial indentation
        String xmlString = "    ";

        //Check if it's a header section
        if(key.trim().equalsIgnoreCase(HEADER_KEY)){
            //Leave a space before it, add the header as a comment
            xmlString = "\n" + xmlString + "<!-- " + string + " -->";
        }
        //If not, treat is as a normal string
        else{
            /* Character checks */
            //Unescaped apostrophes
            string = string.replace("\'", "\\" + "\'");

            //Unescaped @ signs
            string = string.replace("@", "\\" + "@");

            //Ampersands
            if(string.contains("&")){
                string = string.replace("&", "&amp;");
            }

            //Ellipses
            string = string.replace("...", "&#8230;");

            //HTML content
            if(string.contains("<html>") || string.contains("<HTML>")){
                string = string.replace("<html>", "<![CDATA[");
                string = string.replace("</html>", "]]>");
                string = string.replace("<HTML>", "<![CDATA[");
                string = string.replace("</HTML>", "]]>");
            }

            //Add the XML tag
            xmlString = xmlString + "<string name=\"" + key.trim() + "\">" + string + "</string>";
        }

        return xmlString;
    }

    /**
     * Processes the list of parsed Strings into the Android XML document
     *
     * @param language The language to parse the Strings for
     * @throws FileNotFoundException
     * @throws UnsupportedEncodingException
     */
    private static void processAndroidStrings(Language language, List<LanguageString> strings)
            throws FileNotFoundException, UnsupportedEncodingException{

        //Set up the writer for the given language
        PrintWriter writer = new PrintWriter(language.getPath());

        //Add the header
        writer.println(XML_OPENER);
        writer.println(RESOURCES_OPENER);

        //Go through the strings
        for(LanguageString currentString : strings){
            try{
                //If there is no key, we cannot parse it so show a warning and move on
                if(currentString.getKey() == null || currentString.getKey().trim().isEmpty()){
                    System.out.println("Warning: Line " + getLineNumber(currentString, strings) +
                            " has no key, and therefore cannot be parsed");
                    continue;
                }

                //Get the String
                String androidString = getLanguageString(ANDROID, currentString, language);

                //If it is null, there is no value so don't add it
                if(androidString != null){
                    writer.println(androidString);
                }
            }
            catch (Exception e){
                System.out.println("Error on Line " + getLineNumber(currentString, strings));
                e.printStackTrace();
            }
        }

        //Add the resources closing to android files
        writer.println(RESOURCES_CLOSER);

        //Close the writer
        writer.close();
    }

    /* IOS STRING PARSING */

    /**
     * Get the formatted String for the iOS Strings document
     *
     * @param key    The String key
     * @param string The String
     * @return The formatted String
     */
    private static String getIOSString(String key, String string){
        //Replace %s format specifiers with %@
        string = string.replace("%s","%@");
        string = string.replace("$s", "$@");

        //Remove <html> </html>tags
        string = string.replace("<html>", "");
        string = string.replace("</html>", "");
        string = string.replace("<HTML>", "");
        string = string.replace("</HTML>", "");

        //Check if it's a header section
        if(key.equalsIgnoreCase(HEADER_KEY)){
            return "\n" + "/*  " + string + " */";
        }
        //If not, treat is as a normal string
        else{
            return "\"" + key + "\" = \"" + string + "\";";
        }
    }

    /**
     * Processes the list of parsed Strings into the iOS Strings document
     *
     * @param language The language to parse the Strings for
     * @throws FileNotFoundException
     * @throws UnsupportedEncodingException
     */
    public static void processIOSStrings(Language language, List<LanguageString> strings)
            throws FileNotFoundException, UnsupportedEncodingException{

        //Set up the writer for the given language
        PrintWriter writer = new PrintWriter(language.getPath());

        //Go through the strings
        for(LanguageString currentString : strings){
            try{
                //If there is no Id, we cannot parse it so show a warning and continue
                if(currentString.getKey() == null){
                    System.out.println("Warning: Line " + getLineNumber(currentString, strings) +
                            " has no Id, and therefore cannot be parsed");
                    continue;
                }

                //Get the iOS String
                String iOSString = getLanguageString(IOS, currentString, language);

                //If the String is null, there is no value so do not add it
                if(iOSString != null) {
                    writer.println(iOSString);
                }
            }
            catch (Exception e){
                System.out.println("Error on Line " + getLineNumber(currentString, strings));
                e.printStackTrace();
            }
        }

        //Close the writer
        writer.close();
    }
}

File: src/main/kotlin/com/guerinet/sp/StringParser.kt
/*
 * Copyright 2013-2018 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.guerinet.sp

import com.squareup.okhttp.OkHttpClient
import com.squareup.okhttp.Request
import com.squareup.okhttp.Response
import org.supercsv.cellprocessor.ift.CellProcessor
import org.supercsv.io.CsvListReader
import org.supercsv.prefs.CsvPreference
import java.io.BufferedReader
import java.io.FileNotFoundException
import java.io.FileReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.util.*
import java.util.regex.Pattern

/**
 * Main class, executes the main code for parsing the Google Docs file
 * @author Julien Guerinet
 * @since 1.0.0
 */
class StringParser {

    /* INSTANCE VARIABLES */

    /**
     * List of languages we are writing to
     */
    protected var languages: MutableList<Language>

    /**
     * List of Urls we are downloading from
     */
    protected var urls: MutableMap<String, String>

    /**
     * Platform we are downloading for
     */
    protected var platform: String? = null

    /**
     * List of Strings to write
     */
    protected var strings: MutableList<BaseString>

    /**
     * Writer to the current file we are writing to
     */
    protected var writer: PrintWriter

    @JvmStatic
    fun main(args: Array<String>) {
        StringParser().run()
    }

    /**
     * Runs the [StringParser]
     */
    fun run() {
        try {
            setup()
            readFromConfigFile()
            verifyConfigInfo()
            downloadAllStrings()
            verifyKeys()
            writeStrings()
            println("Strings parsing complete")
        } catch (e: IOException) {
            println("Error downloading Strings: ")
            e.printStackTrace()
        }

    }

    /**
     * Sets up the variables
     */
    protected fun setup() {
        // List of all of the languages the Strings are in
        languages = ArrayList()

        // List of Urls to get the Strings from
        urls = HashMap()

        // Platform this is for (null if not chosen)
        platform = null

        // List of Strings to write
        strings = ArrayList()
    }

    /**
     * Reads and parses the various pieces of info from the config file
     *
     * @throws IOException Thrown if there was an error opening or reading the config file
     */
    @Throws(IOException::class)
    protected fun readFromConfigFile() {
        var configReader: BufferedReader? = null
        try {
            configReader = BufferedReader(FileReader("../config.txt"))
        } catch (e: FileNotFoundException) {
            try {
                configReader = BufferedReader(FileReader("config.txt"))
            } catch (ex: FileNotFoundException) {
                println("Error: Config file not found")
                System.exit(-1)
            }

        }

        var line: String
        while ((line = configReader!!.readLine()) != null) {
            readConfigLine(line)
        }
        configReader!!.close()
    }

    protected fun readConfigLine(line: String) {
        if (line.startsWith(URL)) {
            // Get the Url: remove the header and separate the file name from the Url
            val urlString = line.replace(URL, "").trim({ it <= ' ' })
            val urlInfo = urlString.split(",".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()

            if (urlInfo.size < 2) {
                println("Error: The following format has too few " +
                        "arguments for a Url: " + urlString)
                System.exit(-1)
            }

            if (!urlInfo[1].trim({ it <= ' ' }).isEmpty()) {
                // Save it as a new Url in the Url map
                urls[urlInfo[0].trim({ it <= ' ' })] = urlInfo[1].trim({ it <= ' ' })
            }
        } else if (line.startsWith(PLATFORM)) {
            // Get the platform: Remove the header
            val platformString = line.replace(PLATFORM, "").trim({ it <= ' ' })
            if (platformString.equals(ANDROID, ignoreCase = true)) {
                platform = ANDROID
            } else if (platformString.equals(IOS, ignoreCase = true)) {
                platform = IOS
            } else if (platformString.equals(WEB, ignoreCase = true)) {
                platform = WEB
            } else {
                // Not recognized
                println("Error: Platform must be either Android, iOS, or Web.")
                System.exit(-1)
            }
        } else if (line.startsWith(LANGUAGE)) {
            // Get the languages: remove the header and separate the language Id from the path
            val languageString = line.replace(LANGUAGE, "").trim({ it <= ' ' })
            val languageInfo = languageString.split(",".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()

            if (languageInfo.size < 2) {
                println("Error: The following format has too few " +
                        "arguments for a language: " + languageString)
                System.exit(-1)
            }

            // Save it as a new language in the list of languages
            languages.add(Language(languageInfo[0].trim({ it <= ' ' }), languageInfo[1].trim({ it <= ' ' })))
        }
    }

    /**
     * Verifies that all of the config info is present
     */
    protected fun verifyConfigInfo() {
        // Make sure everything is set
        if (urls.isEmpty()) {
            println("Error: There must be at least one non-null Url")
            System.exit(-1)
        } else if (platform == null) {
            println("Error: You need to input a platform")
            System.exit(-1)
        } else if (languages.isEmpty()) {
            println("Error: You need to add at least one language")
            System.exit(-1)
        }

        // Make sure that we have a path per language
        for (language in languages) {
            if (language.path == null) {
                println("Error: Languages need a file path for Android and iOS")
                System.exit(-1)
            }
        }
    }

    /**
     * Downloads all of the Strings from all of the Urls
     *
     * @throws IOException Thrown if there is any error downloading the Strings
     */
    @Throws(IOException::class)
    protected fun downloadAllStrings() {
        for (urlKey in urls.keys) {
            // Go through the Urls and download all of the Strings
            val urlStrings = downloadStrings(urlKey, urls[urlKey])
                    ?: // Don't continue if there's an error downloading the Strings
                    return

            strings.addAll(urlStrings)
        }
    }

    /**
     * Connects to the given Url and downloads the Strings in the right format
     *
     * @param url       Url to connect to
     * @return List of Strings that were downloaded, null if there was an error
     * @throws IOException Thrown if there was any errors downloading the Strings
     */
    @Throws(IOException::class)
    protected fun downloadStrings(urlName: String, url: String): List<BaseString>? {
        // Connect to the URL
        println("Connecting to $url")
        val request = Request.Builder()
                .get()
                .url(url)
                .build()

        val response: Response
        try {
            response = OkHttpClient().setCache(null).newCall(request).execute()
        } catch (e: IOException) {
            // Catch the exception here to be able to continue a build even if we are not connected
            println("IOException while connecting to the URL")
            println("Error Message: " + e.message)
            return null
        }

        val responseCode = response.code()
        println("Response Code: $responseCode")

        if (responseCode != 200) {
            println("Error: Response Message: " + response.message())
            return null
        }

        // Set up the CSV reader
        val reader = CsvListReader(InputStreamReader(response.body().byteStream(),
                "UTF-8"), CsvPreference.EXCEL_PREFERENCE)

        // Keep track of which columns hold the keys and the platform
        var keyColumn = -1
        var platformColumn = -1

        // Get the header
        val header = reader.getHeader(true)

        for (i in header.indices) {
            val string = header[i]
                    ?: // Disregard null headers
                    continue

// Check if the String matches the key key
            if (string.equals(KEY, ignoreCase = true)) {
                keyColumn = i
                continue
            }

            // Check if the String matches the platform key
            if (string.equals(PLATFORMS, ignoreCase = true)) {
                platformColumn = i
                continue
            }

            // Check if the String matches any of the languages parsed
            for (language in languages) {
                if (string.trim { it <= ' ' }.equals(language.id, ignoreCase = true)) {
                    // If we find a match, set the column index for this language
                    language.columnIndex = i
                    break
                }
            }
        }

        // Make sure there is a key column
        if (keyColumn == -1) {
            println("Error: There must be a column marked 'key' with the String keys")
            System.exit(-1)
        }

        // Make sure that all languages have an index
        for (language in languages) {
            if (language.columnIndex == -1) {
                println("Error: " + language.id + " in " + urlName +
                        " does not have any translations.")
                System.exit(-1)
            }
        }

        // Create the list of Strings
        val strings = ArrayList<BaseString>()

        // Make a CellProcessor with the right length
        val processors = arrayOfNulls<CellProcessor>(header.size)

        // Go through each line of the CSV document into a list of objects.
        var currentLine: List<Any>
        // The current line number (start at 2 since 1 is the header)
        var lineNumber = 2
        while ((currentLine = reader.read(*processors)) != null) {
            // Get the key from the current line
            var key: String? = currentLine[keyColumn] as String

            // Check if there's a key
            if (key == null || key.trim { it <= ' ' }.isEmpty()) {
                println("Warning: Line " + lineNumber + " does not have a key and " +
                        "will not be parsed")

                // Increment the line number
                lineNumber++

                // Move on to the new String
                continue
            }

            // Trim the key before continuing
            key = key.trim { it <= ' ' }

            // Check if this is a header
            if (key.startsWith(HEADER_KEY)) {
                strings.add(BaseString(key.replace("###", "").trim { it <= ' ' }, urlName, lineNumber))

                // Increment the line number and continue
                lineNumber++
                continue
            }

            // Add a new language String
            val languageString = LanguageString(key, urlName, lineNumber)

            // Go through the languages, add each translation
            var allNull = true
            var oneNull = false
            for (language in languages) {
                val currentLanguage = currentLine[language.columnIndex] as String

                if (currentLanguage != null) {
                    allNull = false
                } else {
                    oneNull = true
                }

                languageString.addTranslation(language.id, currentLanguage)
            }

            // If there's a platform column , add them
            if (platformColumn != -1) {
                languageString.addPlatforms(currentLine[platformColumn] as String)
            }

            // Check if all of the values are null
            if (allNull) {
                // Show a warning message
                println("Warning: Line " + lineNumber + " from " + urlName +
                        " has no translations so it will not be parsed.")
            } else {
                if (oneNull) {
                    println("Warning: Line " + lineNumber + " from " + urlName +
                            " is missing at least one translation")
                }
                strings.add(languageString)
            }

            // Increment the line number
            lineNumber++
        }

        // Close the CSV reader
        reader.close()

        return strings
    }

    /**
     * Verifies that the keys are valid
     */
    protected fun verifyKeys() {
        // Define the key checker pattern to make sure no illegal characters exist within the keys
        val keyChecker = Pattern.compile("[^A-Za-z0-9_]")

        // Check if there are any errors with the keys
        for (i in strings.indices) {
            val string1 = strings[i] as? LanguageString ?: continue

            // Skip headers for the checks

            // Check if there are any spaces in the keys
            if (string1.key.contains(" ")) {
                println("Error: " + getLog(string1) + " contains a space in its key.")
                System.exit(-1)
            }

            if (keyChecker.matcher(string1.key).find()) {
                println("Error: " + getLog(string1) +
                        " contains some illegal characters.")
                System.exit(-1)
            }

            // Check if there are any duplicates
            for (j in i + 1 until strings.size) {
                val string2 = strings[j]

                // If the keys are the same and it's not a header, show an error and stop
                if (string1.key == string2.key) {
                    println("Error: " + getLog(string1) + " and " + getLog(string2) +
                            " have the same key.")
                    System.exit(-1)
                }
            }
        }
    }

    /**
     * Writes all of the Strings for the different languages
     *
     * @throws IOException Thrown if any error happens
     */
    @Throws(IOException::class)
    protected fun writeStrings() {
        // Go through each language, and write the file
        for (language in languages) {
            // Set up the writer for the given language, enforcing UTF-8
            writer = PrintWriter(language.path, "UTF-8")

            writeStrings(language)

            println("Wrote " + language.id + " to file: " + language.path)

            writer.flush()
            writer.close()
        }
    }

    /**
     * Processes the Strings and writes them to a given file
     *
     * @param language Language we are writing for
     */
    protected fun writeStrings(language: Language) {
        // Header
        writeHeader()

        // Go through the Strings
        for (i in strings.indices) {
            val currentString = strings[i]

            try {
                if (currentString !is LanguageString) {
                    // If we are parsing a header, right the value as a comment
                    writeComment(currentString.key)
                    continue
                } else {
                    writeString(language, currentString, i == strings.size - 1)
                }
            } catch (e: Exception) {
                println("Error on " + getLog(currentString))
                e.printStackTrace()
            }

        }

        // Footer
        writeFooter()
    }

    /**
     * Writes the header to the current file
     */
    protected fun writeHeader() {
        when (platform) {
            ANDROID -> writer.println("<?xml version=\"1.0\" encoding=\"utf-8\"?> \n <resources>")
            WEB -> writer.println("{")
        }
    }

    /**
     * Writes a comment to the file
     *
     * @param comment  Comment String to process
     */
    protected fun writeComment(comment: String) {
        when (platform) {
            ANDROID -> writer.println("\n    <!-- $comment -->")
            IOS -> writer.println("\n/* $comment */")
        }
    }

    /**
     * Writes a String to the file
     *
     * @param language       Current language
     * @param languageString String to write
     * @param lastString     True if this is the last String of the file, false otherwise
     */
    protected fun writeString(language: Language, languageString: LanguageString,
            lastString: Boolean) {
        // Check that it's for the right platform
        if (!languageString.isForPlatform(platform!!)) {
            return
        }

        var string = languageString.getString(language.id)

        // Check if value is or null empty: if it is, continue
        if (string == null) {
            string = ""
        }

        if (string.trim { it <= ' ' }.isEmpty() && !platform!!.equals(WEB, ignoreCase = true)) {
            // Skip over the empty values unless we're on Web
            return
        }

        // Unescaped quotations
        string = string.replace("\"", "\\" + "\"")

        // Copyright
        string = string.replace("(c)", "\u00A9")

        // New Lines
        string = string.replace("\n", "")

        val key = languageString.key
        when (platform) {
            ANDROID -> {
                // Ampersands
                string = string.replace("&", "&amp;")

                // Apostrophes
                string = string.replace("'", "\\'")

                // Unescaped @ signs
                string = string.replace("@", "\\" + "@")

                // Ellipses
                string = string.replace("...", "&#8230;")

                // Check if this is an HTML String
                if (string.contains("<html>") || string.contains("<HTML>")) {
                    // Don't format the greater than and less than symbols
                    string = string.replace("<html>", "<![CDATA[")
                    string = string.replace("</html>", "]]>")
                    string = string.replace("<HTML>", "<![CDATA[")
                    string = string.replace("</HTML>", "]]>")
                } else {
                    // Format the greater then and less than symbol otherwise
                    // Greater than
                    string = string.replace(">", "&gt;")

                    // Less than
                    string = string.replace("<", "&lt;")
                }

                // Add the XML tag
                writer.println("    <string name=\"$key\">$string</string>")
            }
            IOS -> {
                // Replace %s format specifiers with %@
                string = string.replace("%s", "%@")
                string = string.replace("\$s", "$@")

                // Remove <html> </html>tags
                string = string.replace("<html>", "")
                string = string.replace("</html>", "")
                string = string.replace("<HTML>", "")
                string = string.replace("</HTML>", "")

                writer.println("\"$key\" = \"$string\";")
            }
            WEB -> {
                // Remove <html> </html>tags
                string = string.replace("<html>", "")
                string = string.replace("</html>", "")
                string = string.replace("<HTML>", "")
                string = string.replace("</HTML>", "")

                writer.println("    \"" + key + "\": \"" + string + "\"" + if (lastString) "" else ",")
            }
        }
    }

    /**
     * Writes the footer to the current file
     */
    protected fun writeFooter() {
        when (platform) {
            ANDROID -> writer.println("</resources>")
            WEB -> writer.println("}")
        }
    }

    /**
     * @param string String we want to log
     * @return Log message to use to designate the given String
     */
    protected fun getLog(string: BaseString): String {
        return "Line " + string.lineNumber + " from " + string.url
    }

    companion object {

        /** PLATFORM CONSTANTS */
        private const val ANDROID = "Android"
        private const val IOS = "iOS"
        private const val WEB = "Web"

        /* CSV Strings */

        /**
         * Designates which column holds the keys
         */
        private const val KEY = "key"

        /**
         * Designates which column holds the platforms
         */
        private const val PLATFORMS = "platforms"

        /**
         * Designates a header within the Strings document
         */
        private const val HEADER_KEY = "###"
    }
}